// electron-builder/out/targets/squirrelWindows.js
"use strict";

const platformPackager_1 = require("../platformPackager");
const electron_winstaller_fixed_1 = require("electron-winstaller-fixed");
const path = require("path");
const log_1 = require("../log");
const fs_extra_p_1 = require("fs-extra-p");
const repositoryInfo_1 = require("../repositoryInfo");
//noinspection JSUnusedLocalSymbols
const __awaiter = require("../awaiter");
class SquirrelWindowsTarget {
    constructor(packager, appOutDir) {
        this.packager = packager;
        this.appOutDir = appOutDir;
    }
    build(arch) {
        return __awaiter(this, void 0, void 0, function* () {
            const appInfo = this.packager.appInfo;
            const version = appInfo.version;
            const archSuffix = platformPackager_1.getArchSuffix(arch);
            const setupFileName = `${ appInfo.productName } Setup ${ version }${ archSuffix }.exe`;
            const installerOutDir = path.join(this.appOutDir, "..", `win${ platformPackager_1.getArchSuffix(arch) }`);
            yield fs_extra_p_1.emptyDir(installerOutDir);
            const distOptions = yield this.computeEffectiveDistOptions(installerOutDir, setupFileName);
            yield electron_winstaller_fixed_1.createWindowsInstaller(distOptions);
            this.packager.dispatchArtifactCreated(path.join(installerOutDir, setupFileName), `${ appInfo.name }-Setup-${ version }${ archSuffix }.exe`);
            const packagePrefix = `${ appInfo.name }-${ electron_winstaller_fixed_1.convertVersion(version) }-`;
            this.packager.dispatchArtifactCreated(path.join(installerOutDir, `${ packagePrefix }full.nupkg`));
            if (distOptions.remoteReleases != null) {
                this.packager.dispatchArtifactCreated(path.join(installerOutDir, `${ packagePrefix }delta.nupkg`));
            }
            this.packager.dispatchArtifactCreated(path.join(installerOutDir, "RELEASES"));
        });
    }
    computeEffectiveDistOptions(installerOutDir, setupExeName) {
        return __awaiter(this, void 0, void 0, function* () {
            const packager = this.packager;
            let iconUrl = packager.customBuildOptions.iconUrl || packager.devMetadata.build.iconUrl;
            if (iconUrl == null) {
                const info = yield repositoryInfo_1.getRepositoryInfo(packager.metadata, packager.devMetadata);
                if (info != null) {
                    iconUrl = `https://github.com/${ info.user }/${ info.project }/blob/master/${ packager.relativeBuildResourcesDirname }/icon.ico?raw=true`;
                }
                if (iconUrl == null) {
                    throw new Error("iconUrl is not specified, please see https://github.com/electron-userland/electron-builder/wiki/Options#WinBuildOptions-iconUrl");
                }
            }
            checkConflictingOptions(packager.customBuildOptions);
            const appInfo = packager.appInfo;
            const projectUrl = yield appInfo.computePackageUrl();
            const rceditOptions = {
                "version-string": appInfo.versionString,
                "file-version": appInfo.buildVersion,
                "product-version": appInfo.version
            };
            const cscInfo = yield packager.cscInfo;
            const options = Object.assign({
                name: appInfo.name,
                productName: appInfo.productName,
                exe: `${ appInfo.productName }.exe`,
                setupExe: setupExeName,
                title: appInfo.productName,
                appDirectory: this.appOutDir,
                outputDirectory: installerOutDir,
                version: appInfo.version,
                description: appInfo.description,
                authors: appInfo.companyName,
                iconUrl: iconUrl,
                setupIcon: yield packager.iconPath,
                certificateFile: cscInfo == null ? null : cscInfo.file,
                certificatePassword: cscInfo == null ? null : cscInfo.password,
                fixUpPaths: false,
                skipUpdateIcon: true,
                usePackageJson: false,
                extraMetadataSpecs: projectUrl == null ? null : `\n    <projectUrl>${ projectUrl }</projectUrl>`,
                copyright: appInfo.copyright,
                packageCompressionLevel: packager.devMetadata.build.compression === "store" ? 0 : 9,
                sign: {
                    name: appInfo.productName,
                    site: projectUrl,
                    overwrite: true,
                    hash: packager.customBuildOptions.signingHashAlgorithms
                },
                rcedit: rceditOptions
            }, packager.customBuildOptions);
            if (!("loadingGif" in options)) {
                const resourceList = yield packager.resourceList;
                if (resourceList.indexOf("install-spinner.gif") !== -1) {
                    options.loadingGif = path.join(packager.buildResourcesDir, "install-spinner.gif");
                }
            }
            return options;
        });
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SquirrelWindowsTarget;
function checkConflictingOptions(options) {
    for (let name of ["outputDirectory", "appDirectory", "exe", "fixUpPaths", "usePackageJson", "extraFileSpecs", "extraMetadataSpecs", "skipUpdateIcon", "setupExe"]) {
        if (name in options) {
            throw new Error(`Option ${ name } is ignored, do not specify it.`);
        }
    }
    if ("noMsi" in options) {
        log_1.warn(`noMsi is deprecated, please specify as "msi": true if you want to create an MSI installer`);
        options.msi = !options.noMsi;
    }
    const msi = options.msi;
    if (msi != null && typeof msi !== "boolean") {
        throw new Error(`msi expected to be boolean value, but string '"${ msi }"' was specified`);
    }
}
