// electron-builder/out/appInfo.js
"use strict";

const log_1 = require("./log");
const platformPackager_1 = require("./platformPackager");
const util_1 = require("./util");
const repositoryInfo_1 = require("./repositoryInfo");
//noinspection JSUnusedLocalSymbols
const __awaiter = require("./awaiter");
class AppInfo {
    constructor(metadata, devMetadata) {
        this.metadata = metadata;
        this.devMetadata = devMetadata;
        this.description = platformPackager_1.smarten(this.metadata.description);
        // windows-only
        this.versionString = {
            CompanyName: this.companyName,
            FileDescription: this.description,
            ProductName: this.productName,
            InternalName: this.productName,
            LegalCopyright: this.copyright
        };
        let buildVersion = metadata.version;
        this.version = buildVersion;
        const buildNumber = this.buildNumber;
        if (!util_1.isEmptyOrSpaces(buildNumber)) {
            buildVersion += `.${ buildNumber }`;
        }
        this.buildVersion = buildVersion;
    }
    get companyName() {
        return this.metadata.author.name;
    }
    get buildNumber() {
        return this.devMetadata.build["build-version"] || process.env.TRAVIS_BUILD_NUMBER || process.env.APPVEYOR_BUILD_NUMBER || process.env.CIRCLE_BUILD_NUM || process.env.BUILD_NUMBER;
    }
    get id() {
        const appId = this.devMetadata.build["app-bundle-id"];
        if (appId != null) {
            log_1.warn("app-bundle-id is deprecated, please use appId");
        }
        if (this.devMetadata.build.appId != null) {
            return this.devMetadata.build.appId;
        }
        if (appId == null) {
            return `com.electron.${ this.metadata.name.toLowerCase() }`;
        }
        return appId;
    }
    get name() {
        return this.metadata.name;
    }
    get productName() {
        return getProductName(this.metadata, this.devMetadata);
    }
    get copyright() {
        const copyright = this.devMetadata.build["app-copyright"];
        if (copyright != null) {
            return copyright;
        }
        return `Copyright © ${ new Date().getFullYear() } ${ this.metadata.author.name || this.productName }`;
    }
    computePackageUrl() {
        return __awaiter(this, void 0, void 0, function* () {
            const url = this.metadata.homepage || this.devMetadata.homepage;
            if (url != null) {
                return url;
            }
            const info = yield repositoryInfo_1.getRepositoryInfo(this.metadata, this.devMetadata);
            if (info != null) {
                return `https://github.com/${ info.user }/${ info.project }`;
            }
            return null;
        });
    }
}
exports.AppInfo = AppInfo;
function getProductName(metadata, devMetadata) {
    return devMetadata.build.productName || metadata.productName || metadata.name;
}
